piccolo_api/change_password/endpoints.py

```diff
@@ -0,0 +1,181 @@
+from piccolo.apps.user.tables import BaseUser
+from starlette.endpoints import HTTPEndpoint, Request
+from starlette.exceptions import HTTPException
+from starlette.responses import (
+    HTMLResponse,
+    PlainTextResponse,
+    RedirectResponse,
+)
+from starlette.status import HTTP_303_SEE_OTHER
+
+from piccolo_api.session_auth.tables import SessionsBase
+from piccolo_api.shared.auth.styles import Styles
+
+if t.TYPE_CHECKING:  # pragma: no cover
+    from jinja2 import Template
+    from starlette.responses import Response
+
+
+CHANGE_PASSWORD_TEMPLATE_PATH = os.path.join(
+    os.path.dirname(os.path.dirname(__file__)),
+    "templates",
+    "change_password.html",
+)
+
+
+class ChangePasswordEndpoint(HTTPEndpoint, metaclass=ABCMeta):
+    @abstractproperty
+    def _auth_table(self) -> t.Type[BaseUser]:
+        raise NotImplementedError
+
+    @abstractproperty
+    def _change_password_template(self) -> Template:
+        raise NotImplementedError
+
+    @abstractproperty
+    def _styles(self) -> Styles:
+        raise NotImplementedError
+
+    def render_template(
+        self, request: Request, template_context: t.Dict[str, t.Any] = {}
+    ) -> HTMLResponse:
+        # If CSRF middleware is present, we have to include a form field with
+        # the CSRF token. It only works if CSRFMiddleware has
+        # allow_form_param=True, otherwise it only looks for the token in the
+        # header.
+        csrftoken = request.scope.get("csrftoken")
+        csrf_cookie_name = request.scope.get("csrf_cookie_name")
+
+        return HTMLResponse(
+            self._change_password_template.render(
+                csrftoken=csrftoken,
+                csrf_cookie_name=csrf_cookie_name,
+                request=request,
+                styles=self._styles,
+                **template_context,
+            )
+        )
+
+    async def get(
+        self, request: Request
+    ) -> t.Union[HTMLResponse, PlainTextResponse]:
+        session_user = (
+            await SessionsBase.select(SessionsBase.user_id)
+            .where(SessionsBase.token == request.cookies.get("id"))
+            .first()
+        )
+        if session_user:
+            return self.render_template(request)
+        else:
+            return PlainTextResponse("Not authenticated", status_code=401)
+
+    async def post(self, request: Request) -> Response:
+        # Some middleware (for example CSRF) has already awaited the request
+        # body, and adds it to the request.
+        body = request.scope.get("form")
+
+        if not body:
+            try:
+                body = await request.json()
+            except JSONDecodeError:
+                body = await request.form()
+
+        old_password = body.get("old_password", None)
+        new_password = body.get("new_password", None)
+        confirm_password = body.get("confirm_password", None)
+
+        if (not old_password) or (not new_password) or (not confirm_password):

```

Comment:
We also need to check that the old password is correct.

For example:

```
piccolo_user = request.user.user
if not await piccolo_user.__class__.login(username=piccolo_user.username, password=old_password):
    raise HTTPException(status_code=401, detail="Incorrect password")
```

Reply:
You're right. I will add that.